### PR TITLE
 [FIX] web_editor: do not show overlay for invisible elements

### PR DESCRIPTION
**Pull request by ThanhDodeurOdoo**
When a snippet is out of the viewport or hidden, the overlay should
not be visible. Two distinct cases we had:

- When the page is scrolled, the header is sometimes hidden thanks to
  tranform: translate(0, -100%); for some kind of effect -> the header
  is thus still visible but out of the viewport, any selected snippet
  inside that header would still have this overlay enabled so we would
  see the move and delete button on the top of the screen.

- When the page is scrolled, with the vertical header, some part of the
  header are hidden for the affixed version. If any snippet in those
  hidden elements was selected, the overlay is still visible floating.

task-2312878

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 1]
```diff
@@ -446,6 +447,17 @@ var SnippetEditor = Widget.extend({
...
+     * @private
+     * @param {DOMElement} element
+     */
+    _hasArea: function (element) {
+        const rect = element.getBoundingClientRect();
+        return (
+             rect.bottom - rect.top > 0 &&
+             rect.right - rect.left > 0
+        );
+    },
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:460]: `.is(':visible')` above already does that, why is this necessary? cc @bvr-odoo

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 2]
```diff
@@ -346,6 +346,9 @@ var SnippetEditor = Widget.extend({
 
         // Show/hide overlay in preview mode or not
         this.$el.toggleClass('oe_active', show);
+        const $optionWrap = this.$el.find('.o_overlay_options_wrap');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:349]: Can we avoid to search the DOM for this each time the overlay is shown/hidden (especially since not needed when hidden) ?

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 3]
```diff
@@ -346,6 +346,9 @@ var SnippetEditor = Widget.extend({
...
+        const isAtPageTop = show && $optionWrap.length && this.$target.get(0).getBoundingClientRect().top + window.scrollY < 20;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:350]: window.scrollY is always 0 since it is never the window which scrolls... something is wrong here

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 4]
```diff
@@ -399,6 +402,9 @@ var SnippetEditor = Widget.extend({
      */
     toggleOverlayVisibility: function (show) {
         if (this.$el && !this.scrollingTimeout) {
+            if (!this._isInViewport(this.$target.get(0))) {
+                show = false;
+            }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js]: I am not sure to get the point of this code, can you explain? We may enabled snippets out of viewport via code then if the user scroll, the overlay should be ok... right?

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 3 above)
Comment (ThanhDodeurOdoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:350]: Indeed, this part is useless

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 4 above)
Comment (ThanhDodeurOdoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: @bvr-odoo found out that sometimes snippets are "hidden" out of the viewport, which is now noticeable with my fix for the handle buttons, since when a snippet is hidden above the viewport we now see the handles dangling from above.

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 1 above)
Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:459]: @qsm-odoo It's for when an element size is 0px * 0px (like the logo image with some header templates). `.is(':visible')` is not enough to hide the overlay in this case

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:459]: https://api.jquery.com/visible-selector/
> Elements are considered visible if they consume space in the document. Visible elements have a width or height that is greater than zero.

Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:459]: @qsm-odoo Indeed but Im testing the code and `.is(':visible')` return 'true' even if the img of the logo is 0px X 0px.
I am investigating

Comment (bvr-odoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:459]: @qsm-odoo I dont clearly understand why but it's always visible. I changed opacity, remove img src, etc. It doesnt change..

**Review by qsm-odoo — APPROVED** [review]
@robodoo r+
